// This file is a part of MapSplice3. Please refer to LICENSE.TXT for the LICENSE
//chr1    lincRNA gene    29554   31109   .       +       .       gene_id "ENSG00000243485"; gene_name "MIR1302-10"; gene_source "ensembl_havana"; gene_biotype "lincRNA";
//chr1    lincRNA transcript      29554   31097   .       +       .       gene_id "ENSG00000243485"; transcript_id "ENST00000473358"; gene_name "MIR1302-10"; gene_source "ensembl_havana"; gene_biotype "lincRNA"; transcript_name "MIR1302-10-001"; transcript_source "havana";
//chr1    lincRNA exon    29554   30039   .       +       .       gene_id "ENSG00000243485"; transcript_id "ENST00000473358"; exon_number "1"; gene_name "MIR1302-10"; gene_source "ensembl_havana"; gene_biotype "lincRNA"; transcript_name "MIR1302-10-001"; transcript_source "havana"; exon_id "ENSE
//00001947070";
//chr1    lincRNA exon    30564   30667   .       +       .       gene_id "ENSG00000243485"; transcript_id "ENST00000473358"; exon_number "2"; gene_name "MIR1302-10"; gene_source "ensembl_havana"; gene_biotype "lincRNA"; transcript_name "MIR1302-10-001"; transcript_source "havana"; exon_id "ENSE
//00001922571";
//chr1    lincRNA exon    30976   31097   .       +       .       gene_id "ENSG00000243485"; transcript_id "ENST00000473358"; exon_number "3"; gene_name "MIR1302-10"; gene_source "ensembl_havana"; gene_biotype "lincRNA"; transcript_name "MIR1302-10-001"; transcript_source "havana"; exon_id "ENSE
//00001827679";
//chr1    lincRNA transcript      30267   31109   .       +       .       gene_id "ENSG00000243485"; transcript_id "ENST00000469289"; gene_name "MIR1302-10"; gene_source "ensembl_havana"; gene_biotype "lincRNA"; transcript_name "MIR1302-10-002"; transcript_source "havana";
//chr1    lincRNA exon    30267   30667   .       +       .       gene_id "ENSG00000243485"; transcript_id "ENST00000469289"; exon_number "1"; gene_name "MIR1302-10"; gene_source "ensembl_havana"; gene_biotype "lincRNA"; transcript_name "MIR1302-10-002"; transcript_source "havana"; exon_id "ENSE
//00001841699";
//chr1    lincRNA exon    30976   31109   .       +       .       gene_id "ENSG00000243485"; transcript_id "ENST00000469289"; exon_number "2"; gene_name "MIR1302-10"; gene_source "ensembl_havana"; gene_biotype "lincRNA"; transcript_name "MIR1302-10-002"; transcript_source "havana"; exon_id "ENSE
//00001890064";
//chr1    miRNA   transcript      30366   30503   .       +       .       gene_id "ENSG00000243485"; transcript_id "ENST00000607096"; gene_name "MIR1302-10"; gene_source "ensembl_havana"; gene_biotype "lincRNA"; transcript_name "MIR1302-10-201"; transcript_source "ensembl";
//chr1    miRNA   exon    30366   30503   .       +       .       gene_id "ENSG00000243485"; transcript_id "ENST00000607096"; exon_number "1"; gene_name "MIR1302-10"; gene_source "ensembl_havana"; gene_biotype "lincRNA"; transcript_name "MIR1302-10-201"; transcript_source "ensembl"; exon_id "ENS
//E00003695741";
//chr1    lincRNA gene    34554   36081   .       -       .       gene_id "ENSG00000237613"; gene_name "FAM138A"; gene_source "ensembl_havana"; gene_biotype "lincRNA";
//chr1    lincRNA transcript      34554   36081   .       -       .       gene_id "ENSG00000237613"; transcript_id "ENST00000417324"; gene_name "FAM138A"; gene_source "ensembl_havana"; gene_biotype "lincRNA"; transcript_name "FAM138A-001"; transcript_source "ensembl_havana";
//chr1    lincRNA exon    35721   36081   .       -       .       gene_id "ENSG00000237613"; transcript_id "ENST00000417324"; exon_number "1"; gene_name "FAM138A"; gene_source "ensembl_havana"; gene_biotype "lincRNA"; transcript_name "FAM138A-001"; transcript_source "ensembl_havana"; exon_id "EN
//SE00001656588";
//chr1    lincRNA exon    35277   35481   .       -       .       gene_id "ENSG00000237613"; transcript_id "ENST00000417324"; exon_number "2"; gene_name "FAM138A"; gene_source "ensembl_havana"; gene_biotype "lincRNA"; transcript_name "FAM138A-001"; transcript_source "ensembl_havana"; exon_id "EN
//SE00001669267";
//chr1    lincRNA exon    34554   35174   .       -       .       gene_id "ENSG00000237613"; transcript_id "ENST00000417324"; exon_number "3"; gene_name "FAM138A"; gene_source "ensembl_havana"; gene_biotype "lincRNA"; transcript_name "FAM138A-001"; transcript_source "ensembl_havana"; exon_id "EN
//SE00001727627";
//chr1    lincRNA transcript      35245   36073   .       -       .       gene_id "ENSG00000237613"; transcript_id "ENST00000461467"; gene_name "FAM138A"; gene_source "ensembl_havana"; gene_biotype "lincRNA"; transcript_name "FAM138A-002"; transcript_source "havana";
//chr1    lincRNA exon    35721   36073   .       -       .       gene_id "ENSG00000237613"; transcript_id "ENST00000461467"; exon_number "1"; gene_name "FAM138A"; gene_source "ensembl_havana"; gene_biotype "lincRNA"; transcript_name "FAM138A-002"; transcript_source "havana"; exon_id "ENSE000016
//18781";
//chr1    lincRNA exon    35245   35481   .       -       .       gene_id "ENSG00000237613"; transcript_id "ENST00000461467"; exon_number "2"; gene_name "FAM138A"; gene_source "ensembl_havana"; gene_biotype "lincRNA"; transcript_name "FAM138A-002"; transcript_source "havana"; exon_id "ENSE000018
//74421";
//chr1    pseudogene      gene    52473   54936   .       +       .       gene_id "ENSG00000268020"; gene_name "OR4G4P"; gene_source "ensembl_havana"; gene_biotype "pseudogene";
//chr1    unprocessed_pseudogene  transcript      52473   53312   .       +       .       gene_id "ENSG00000268020"; transcript_id "ENST00000606857"; gene_name "OR4G4P"; gene_source "ensembl_havana"; gene_biotype "pseudogene"; transcript_name "OR4G4P-001"; transcript_source "havana";
//chr1    unprocessed_pseudogene  exon    52473   53312   .       +       .       gene_id "ENSG00000268020"; transcript_id "ENST00000606857"; exon_number "1"; gene_name "OR4G4P"; gene_source "ensembl_havana"; gene_biotype "pseudogene"; transcript_name "OR4G4P-001"; transcript_source "havana"; ex
//on_id "ENSE00003698237";
//chr1    unprocessed_pseudogene  transcript      53049   54936   .       +       .       gene_id "ENSG00000268020"; transcript_id "ENST00000594647"; gene_name "OR4G4P"; gene_source "ensembl_havana"; gene_biotype "pseudogene"; transcript_name "OR4G4P-201"; transcript_source "ensembl";
//chr1    unprocessed_pseudogene  exon    53049   53067   .       +       .       gene_id "ENSG00000268020"; transcript_id "ENST00000594647"; exon_number "1"; gene_name "OR4G4P"; gene_source "ensembl_havana"; gene_biotype "pseudogene"; transcript_name "OR4G4P-201"; transcript_source "ensembl"; e
//xon_id "ENSE00003076518";
//chr1    unprocessed_pseudogene  exon    54830   54936   .       +       .       gene_id "ENSG00000268020"; transcript_id "ENST00000594647"; exon_number "2"; gene_name "OR4G4P"; gene_source "ensembl_havana"; gene_biotype "pseudogene"; transcript_name "OR4G4P-201"; transcript_source "ensembl"; e
//xon_id "ENSE00003074125";
//chr1    pseudogene      gene    62948   63887   .       +       .       gene_id "ENSG00000240361"; gene_name "OR4G11P"; gene_source "havana"; gene_biotype "pseudogene";
//chr1    unprocessed_pseudogene  transcript      62948   63887   .       +       .       gene_id "ENSG00000240361"; transcript_id "ENST00000492842"; gene_name "OR4G11P"; gene_source "havana"; gene_biotype "pseudogene"; transcript_name "OR4G11P-001"; transcript_source "havana";
//chr1    unprocessed_pseudogene  exon    62948   63887   .       +       .       gene_id "ENSG00000240361"; transcript_id "ENST00000492842"; exon_number "1"; gene_name "OR4G11P"; gene_source "havana"; gene_biotype "pseudogene"; transcript_name "OR4G11P-001"; transcript_source "havana"; exon_id
//"ENSE00001830178";
//chr1    protein_coding  gene    69091   70008   .       +       .       gene_id "ENSG00000186092"; gene_name "OR4F5"; gene_source "ensembl_havana"; gene_biotype "protein_coding";
//chr1    protein_coding  transcript      69091   70008   .       +       .       gene_id "ENSG00000186092"; transcript_id "ENST00000335137"; gene_name "OR4F5"; gene_source "ensembl_havana"; gene_biotype "protein_coding"; transcript_name "OR4F5-001"; transcript_source "ensembl_havana"; tag "CCDS
//"; ccds_id "CCDS30547";
//chr1    protein_coding  exon    69091   70008   .       +       .       gene_id "ENSG00000186092"; transcript_id "ENST00000335137"; exon_number "1"; gene_name "OR4F5"; gene_source "ensembl_havana"; gene_biotype "protein_coding"; transcript_name "OR4F5-001"; transcript_source "ensembl_havana";
//tag "CCDS"; ccds_id "CCDS30547"; exon_id "ENSE00002319515";
//chr1    protein_coding  CDS     69091   70005   .       +       0       gene_id "ENSG00000186092"; transcript_id "ENST00000335137"; exon_number "1"; gene_name "OR4F5"; gene_source "ensembl_havana"; gene_biotype "protein_coding"; transcript_name "OR4F5-001"; transcript_source "ensembl_havana";
//tag "CCDS"; ccds_id "CCDS30547"; protein_id "ENSP00000334393";

#ifndef LOTHELLOGENECOUNT_GENE_INFO_H
#define LOTHELLOGENECOUNT_GENE_INFO_H

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <malloc.h>
#include <string.h>
#include <iostream>
#include <fstream>
#include <stack>
#include <vector>
#include <hash_map>
#include <map>
#include <set>
#include <sstream>
#include "../../../../general/index_info.h"
using namespace std;

class Gene_Transcript_Info
{
private:
	string geneName;
	string geneId;
	int chrNameInt;
	int startPos;
	int endPos;
	vector<string> transcriptIdVec;
	vector<string> transcriptNameVec;
	vector< vector< pair<int,int> > > transcriptExonicPosPairVecVec;
	//vector< pair<int,int> > exonicPosPairVec;
	//vector< pair<int,int> > intronicPosPairVec;
public:
	Gene_Transcript_Info()
	{}

	void initiate(string& tmpGeneChrName, string& tmpGeneName, string& tmpGeneId, 
		vector<string>& tmpTranscriptNameVec, vector<string>& tmpTranscriptIdVec, 
		vector< vector< pair<int,int> > >& tmpExonPosPairVecVec, Index_Info* indexInfo)
	{
		chrNameInt = indexInfo->convertStringToInt(tmpGeneChrName);
		geneName = tmpGeneName;
		geneId = tmpGeneId;
		for(int tmp = 0; tmp < tmpTranscriptNameVec.size(); tmp++)
		{
			transcriptNameVec.push_back(tmpTranscriptNameVec[tmp]);
			transcriptIdVec.push_back(tmpTranscriptIdVec[tmp]);
			transcriptExonicPosPairVecVec.push_back(tmpExonPosPairVecVec[tmp]);
		}
		vector<int> tmpGeneExonStartPosVec;
		vector<int> tmpGeneExonEndPosVec;
		for(int tmp1 = 0; tmp1 < transcriptExonicPosPairVecVec.size(); tmp1++)
		{
			for(int tmp2 = 0; tmp2 < (transcriptExonicPosPairVecVec[tmp1]).size(); tmp2++)
			{
				int tmpExonStartPos = (transcriptExonicPosPairVecVec[tmp1])[tmp2].first;
				int tmpExonEndPos = (transcriptExonicPosPairVecVec[tmp1])[tmp2].second;
				tmpGeneExonStartPosVec.push_back(tmpExonStartPos);
				tmpGeneExonEndPosVec.push_back(tmpExonEndPos);
			}
		}
		// get startPos
		int tmpSmallestStartPos = tmpGeneExonStartPosVec[0];
		if(tmpGeneExonStartPosVec.size() > 1)
		{
			for(int tmp = 1; tmp < tmpGeneExonStartPosVec.size(); tmp++)
			{
				int tmpStartPos = tmpGeneExonStartPosVec[tmp];
				if(tmpStartPos < tmpSmallestStartPos)
					tmpSmallestStartPos = tmpStartPos;
			}
		}
		startPos = tmpSmallestStartPos;
		// get endPos
		int tmpLargestEndPos = tmpGeneExonEndPosVec[0];
		if(tmpGeneExonEndPosVec.size() > 1)
		{
			for(int tmp = 1; tmp < tmpGeneExonEndPosVec.size(); tmp++)
			{
				int tmpEndPos = tmpGeneExonEndPosVec[tmp];
				if(tmpEndPos > tmpLargestEndPos)
					tmpLargestEndPos = tmpEndPos;
			}
		}
		endPos = tmpLargestEndPos;
	}

	string return_geneName()
	{
		return geneName;
	}

	int returnChrNameInt()
	{
		return chrNameInt;
	}

	int returnStartPos()
	{
		return startPos;
	}

	int returnEndPos()
	{
		return endPos;
	}

	int returnTranscriptIsoformNum()
	{
		return transcriptExonicPosPairVecVec.size();
	}

	string returnFa_gene(Index_Info* indexInfo)
	{
		string tmpId = ">Gene_" + geneName + "_" + geneId;

		vector<string> tmpSeqVec; //= "";
		int geneLength = endPos - startPos + 1;
		int fiftyBaseLineNum = geneLength/50;
		for(int tmp = 0; tmp < fiftyBaseLineNum; tmp++)
			tmpSeqVec.push_back(indexInfo->returnChromStrSubstr(chrNameInt, startPos + tmp * 50, 50));
		if(fiftyBaseLineNum * 50 < geneLength)
		{
			int lastLineLength = geneLength - fiftyBaseLineNum * 50;
			tmpSeqVec.push_back(indexInfo->returnChromStrSubstr(chrNameInt, 
				startPos + fiftyBaseLineNum * 50, lastLineLength));
		}

		string tmpFaStr = tmpId;
		for(int tmp = 0; tmp < tmpSeqVec.size(); tmp++)
		{
			tmpFaStr += "\n";
			tmpFaStr += tmpSeqVec[tmp];
		}
		return tmpFaStr;
	}

	string returnFa_transcript(Index_Info* indexInfo, int isoformIndex)
	{
		string tmpTranscriptName = transcriptNameVec[isoformIndex];
		string tmpTranscriptId = transcriptIdVec[isoformIndex];
		string tmpId = ">Transcript_" + tmpTranscriptName + "_" + tmpTranscriptId;
		vector<int> tmpStartPosVec;
		vector<int> tmpEndPosVec;
		for(int tmp = 0; tmp < (transcriptExonicPosPairVecVec[isoformIndex]).size(); tmp++)
		{
			tmpStartPosVec.push_back((transcriptExonicPosPairVecVec[isoformIndex])[tmp].first);
			tmpEndPosVec.push_back((transcriptExonicPosPairVecVec[isoformIndex])[tmp].second);
		}
		// sort startPosVec, endPosVec
		sort(tmpStartPosVec.begin(), tmpStartPosVec.end());
		sort(tmpEndPosVec.begin(), tmpEndPosVec.end());
		
		string tmpTranscriptSeq = "";
		int tmpExonNum = tmpStartPosVec.size();
		for(int tmp = 0; tmp < tmpExonNum; tmp++)
		{
			int tmpStartPos = tmpStartPosVec[tmp];
			int tmpEndPos = tmpEndPosVec[tmp];
			string tmpExonSeq = indexInfo->returnChromStrSubstr(chrNameInt, 
				tmpStartPos, tmpEndPos - tmpStartPos + 1);
			tmpTranscriptSeq += tmpExonSeq;
		}
	
		vector<string> tmpSeqVec;
		int tmpTranscriptSeqLength = tmpTranscriptSeq.length();
		int fiftyBaseLineNum = tmpTranscriptSeqLength/50;
		for(int tmp = 0; tmp < fiftyBaseLineNum; tmp++)
			tmpSeqVec.push_back(tmpTranscriptSeq.substr(tmp * 50, 50));
		if(fiftyBaseLineNum * 50 < tmpTranscriptSeqLength)
		{
			int lastLineLength = tmpTranscriptSeqLength - fiftyBaseLineNum * 50;
			tmpSeqVec.push_back(tmpTranscriptSeq.substr(fiftyBaseLineNum * 50, lastLineLength));
		}

		string tmpFaStr = tmpId;
		for(int tmp = 0; tmp < tmpSeqVec.size(); tmp++)
		{
			tmpFaStr += "\n";
			tmpFaStr += tmpSeqVec[tmp];
		}
		return tmpFaStr;		
	}

	string returnFa_transcriptVec(Index_Info* indexInfo)
	{
		if(transcriptExonicPosPairVecVec.size() == 0)
		{
			cout << "error ! transcriptExonicPosPairVecVec.size() == 0" << endl;
			exit(1);
		}
		string tmpStr = this->returnFa_transcript(indexInfo, 0);
		if(transcriptExonicPosPairVecVec.size() > 1)
		{	
			for(int tmp = 0; tmp < transcriptExonicPosPairVecVec.size(); tmp++)
			{
				string tmpFa_transcript = this->returnFa_transcript(indexInfo, tmp);
				tmpStr += "\n";
				tmpStr += tmpFa_transcript;
			}
		}
		return tmpStr;
	}

	void printoutFa_gene(string& tmpFaFile, Index_Info* indexInfo)
	{
		ofstream fa_ofs(tmpFaFile.c_str());
		string tmpFa_gene = this->returnFa_gene(indexInfo);
		fa_ofs << tmpFa_gene << endl;
		fa_ofs.close();
	}

	void printoutFa_transcript(string& tmpFaFile, Index_Info* indexInfo)
	{
		ofstream fa_ofs(tmpFaFile.c_str());
		string tmpFa_transcript = this->returnFa_transcriptVec(indexInfo);
		fa_ofs << tmpFa_transcript << endl;
		fa_ofs.close();
	}

	void printoutFa_gene_transcript(string& tmpFaFile, Index_Info* indexInfo)
	{
		ofstream fa_ofs(tmpFaFile.c_str());
		string tmpFa_gene = this->returnFa_gene(indexInfo);
		string tmpFa_transcript = this->returnFa_transcriptVec(indexInfo);
		fa_ofs << tmpFa_gene << endl << tmpFa_transcript << endl;
		fa_ofs.close();
	}
};

class Gene_Transcript_Vec_Info 
{
private:
	vector<Gene_Transcript_Info> geneTranscriptInfoVec;

public:
	Gene_Transcript_Vec_Info()
	{}

	void initaite_ensemblGeneAnn(string& ensemblGeneAnnFile, Index_Info* indexInfo)
	{
		vector<string> chrNameVec_exon;
		vector<string> featureVec_exon;
		vector<int> startPosVec_exon;
		vector<int> endPosVec_exon;
		vector<string> strandVec_exon;
		vector<string> transcriptIdVec_exon;
		vector<string> transcriptNameVec_exon;
		vector<string> geneIdVec_exon;
		vector<string> geneNameVec_exon;		
		ifstream ann_ifs(ensemblGeneAnnFile.c_str());
		int tmpAnnLineNO = 0;
		cout << "start to generate exonVec from annStrVec" << endl;
		while(!ann_ifs.eof())
		{
			string tmpStr;
			getline(ann_ifs, tmpStr);
			tmpAnnLineNO ++;
			int tmpThousandIndex = tmpAnnLineNO / 100000;
			if(tmpAnnLineNO == tmpThousandIndex * 100000)
			cout << "Tmp Ann #: " << tmpAnnLineNO << endl;
			if(tmpStr == "")
				break;
			//cout << "tmpStr: " << tmpStr << endl;
			string tmpChrName, tmpFeature, tmpStrand, tmpTranscriptId, tmpTranscriptName, tmpGeneId, tmpGeneName;
			int tmpStartPos, tmpEndPos;
			bool tmpParseEnsemblStr_bool = parseEnsemblStr(tmpStr, tmpChrName, tmpFeature, tmpStartPos, tmpEndPos, tmpStrand, tmpTranscriptId, tmpTranscriptName, tmpGeneId, tmpGeneName);
			//cout << "tmpParseEnsemblStr_bool: " << tmpParseEnsemblStr_bool << endl;
			if(tmpParseEnsemblStr_bool)
			{	
				//cout << "tmpChrName: " << tmpChrName << endl;
				int tmpChrNameInt = indexInfo->convertStringToInt(tmpChrName);
				//cout << "tmpChrNameInt: " << tmpChrNameInt << endl;
				if(tmpChrNameInt < 0)
					continue;
				chrNameVec_exon.push_back(tmpChrName);
				featureVec_exon.push_back(tmpFeature);
				startPosVec_exon.push_back(tmpStartPos);
				endPosVec_exon.push_back(tmpEndPos);
				strandVec_exon.push_back(tmpStrand);
				transcriptIdVec_exon.push_back(tmpTranscriptId);
				transcriptNameVec_exon.push_back(tmpTranscriptName);
				geneIdVec_exon.push_back(tmpGeneId);
				geneNameVec_exon.push_back(tmpGeneName);
			}
		}
		ann_ifs.close();
		cout << "exon #: " << chrNameVec_exon.size() << endl;
		vector<string> geneChrNameVec;
		vector<string> geneNameVec;
		vector<string> geneIdVec;
		vector< vector<string> > transcriptNameVecVec;
		vector< vector<string> > transcriptIdVecVec;
		vector< vector< vector< pair<int,int> > > > exonPosPairVecVecVec;
		int exonNum = chrNameVec_exon.size();		
		//generate geneNameVec, geneIdVec, transcriptNameVecVec, transcriptIdVecVec, exonPosPairVecVecVec
		int tmpExonLineNO = 0;
		cout << "start to generate geneVec from exonVec" << endl;
		for(int tmp = 0; tmp < exonNum; tmp++)
		{
			//cout << "tmpExonNum: " << tmp << endl;
			tmpExonLineNO ++;
			int tmpThousandIndex = tmpExonLineNO / 10000;
			if(tmpExonLineNO == tmpThousandIndex * 10000)
				cout << "Tmp Exon #: " << tmpExonLineNO << endl;

			string tmpChrName = chrNameVec_exon[tmp];
			string tmpFeature = featureVec_exon[tmp];
			int tmpStartPos = startPosVec_exon[tmp];
			int tmpEndPos = endPosVec_exon[tmp];
			string tmpStrand = strandVec_exon[tmp];
			string tmpTranscriptId = transcriptIdVec_exon[tmp];
			string tmpTranscriptName = transcriptNameVec_exon[tmp];
			string tmpGeneId = geneIdVec_exon[tmp];
			string tmpGeneName = geneNameVec_exon[tmp];
			
			int currentGeneNameVecSize = geneNameVec.size();
			string currentLastGeneName;
			if(currentGeneNameVecSize == 0)
				currentLastGeneName = ""; 
			else
				currentLastGeneName = geneNameVec[currentGeneNameVecSize - 1];
			if(tmpGeneName == currentLastGeneName) // existing gene
			{
				//cout << "existing gene" << endl;
				int currentTranscriptNameVecSize = transcriptNameVecVec[currentGeneNameVecSize - 1].size();
				string currentLastGeneLastTranscriptName;
				if(currentTranscriptNameVecSize == 0)
					currentLastGeneLastTranscriptName = "";
				else
					currentLastGeneLastTranscriptName = (transcriptNameVecVec[currentGeneNameVecSize - 1])[currentTranscriptNameVecSize - 1];
				if(tmpTranscriptName == currentLastGeneLastTranscriptName)
					((exonPosPairVecVecVec[currentGeneNameVecSize - 1])[currentTranscriptNameVecSize - 1]).push_back(pair<int,int>(tmpStartPos, tmpEndPos));
				else // new transcript
				{
					(transcriptNameVecVec[currentGeneNameVecSize - 1]).push_back(tmpTranscriptName);
					(transcriptIdVecVec[currentGeneNameVecSize - 1]).push_back(tmpTranscriptId);
					vector< pair<int,int> > tmpNewTranscriptExonPosPairVec;
					tmpNewTranscriptExonPosPairVec.push_back(pair<int,int>(tmpStartPos, tmpEndPos));
					(exonPosPairVecVecVec[currentGeneNameVecSize - 1]).push_back(tmpNewTranscriptExonPosPairVec);
				}
			}
			else // new gene
			{
				//cout << "new gene" << endl;
				geneChrNameVec.push_back(tmpChrName);
				geneNameVec.push_back(tmpGeneName);
				geneIdVec.push_back(tmpGeneId);
				
				vector<string> tmpNewTranscriptNameVec;
				tmpNewTranscriptNameVec.push_back(tmpTranscriptName);
				transcriptNameVecVec.push_back(tmpNewTranscriptNameVec);

				vector<string> tmpNewTranscriptIdVec;
				tmpNewTranscriptIdVec.push_back(tmpTranscriptId);
				transcriptIdVecVec.push_back(tmpNewTranscriptIdVec);

				vector< pair<int,int> > tmpExonPosPairVec;
				tmpExonPosPairVec.push_back(pair<int,int>(tmpStartPos, tmpEndPos));
				vector< vector< pair<int,int> > > tmpNewTranscriptExonPosPairVecVec;
				tmpNewTranscriptExonPosPairVecVec.push_back(tmpExonPosPairVec);
				exonPosPairVecVecVec.push_back(tmpNewTranscriptExonPosPairVecVec);				
			}
		}
		cout << "gene #: " << geneNameVec.size() << endl;
		for(int tmp = 0; tmp < geneNameVec.size(); tmp++)
		{
			Gene_Transcript_Info tmpGeneTranscriptInfo;
			tmpGeneTranscriptInfo.initiate(geneChrNameVec[tmp], geneNameVec[tmp], geneIdVec[tmp], 
				transcriptNameVecVec[tmp], transcriptIdVecVec[tmp], exonPosPairVecVecVec[tmp], indexInfo);
			geneTranscriptInfoVec.push_back(tmpGeneTranscriptInfo);
		}
	}

	void initaite_ensemblGeneAnn_old(string& ensemblGeneAnnFile, Index_Info* indexInfo)
	{
		vector<string> chrNameVec_exon;
		vector<string> featureVec_exon;
		vector<int> startPosVec_exon;
		vector<int> endPosVec_exon;
		vector<string> strandVec_exon;
		vector<string> transcriptIdVec_exon;
		vector<string> transcriptNameVec_exon;
		vector<string> geneIdVec_exon;
		vector<string> geneNameVec_exon;		
		ifstream ann_ifs(ensemblGeneAnnFile.c_str());
		int tmpAnnLineNO = 0;
		cout << "start to generate exonVec from annStrVec" << endl;
		while(!ann_ifs.eof())
		{
			string tmpStr;
			getline(ann_ifs, tmpStr);
			tmpAnnLineNO ++;
			int tmpThousandIndex = tmpAnnLineNO / 100000;
			if(tmpAnnLineNO == tmpThousandIndex * 100000)
			cout << "Tmp Ann #: " << tmpAnnLineNO << endl;
			if(tmpStr == "")
				break;
			//cout << "tmpStr: " << tmpStr << endl;
			string tmpChrName, tmpFeature, tmpStrand, tmpTranscriptId, tmpTranscriptName, tmpGeneId, tmpGeneName;
			int tmpStartPos, tmpEndPos;
			bool tmpParseEnsemblStr_bool = parseEnsemblStr(tmpStr, tmpChrName, tmpFeature, tmpStartPos, tmpEndPos, tmpStrand, tmpTranscriptId, tmpTranscriptName, tmpGeneId, tmpGeneName);
			//cout << "tmpParseEnsemblStr_bool: " << tmpParseEnsemblStr_bool << endl;
			if(tmpParseEnsemblStr_bool)
			{	
				//cout << "tmpChrName: " << tmpChrName << endl;
				int tmpChrNameInt = indexInfo->convertStringToInt(tmpChrName);
				//cout << "tmpChrNameInt: " << tmpChrNameInt << endl;
				if(tmpChrNameInt < 0)
					continue;
				chrNameVec_exon.push_back(tmpChrName);
				featureVec_exon.push_back(tmpFeature);
				startPosVec_exon.push_back(tmpStartPos);
				endPosVec_exon.push_back(tmpEndPos);
				strandVec_exon.push_back(tmpStrand);
				transcriptIdVec_exon.push_back(tmpTranscriptId);
				transcriptNameVec_exon.push_back(tmpTranscriptName);
				geneIdVec_exon.push_back(tmpGeneId);
				geneNameVec_exon.push_back(tmpGeneName);
			}
		}
		ann_ifs.close();
		cout << "exon #: " << chrNameVec_exon.size() << endl;
		vector<string> geneChrNameVec;
		vector<string> geneNameVec;
		vector<string> geneIdVec;
		vector< vector<string> > transcriptNameVecVec;
		vector< vector<string> > transcriptIdVecVec;
		vector< vector< vector< pair<int,int> > > > exonPosPairVecVecVec;
		int exonNum = chrNameVec_exon.size();
		
		//generate geneNameVec, geneIdVec, transcriptNameVecVec, transcriptIdVecVec, exonPosPairVecVecVec
		int tmpExonLineNO = 0;
		cout << "start to generate geneVec from exonVec" << endl;
		for(int tmp = 0; tmp < exonNum; tmp++)
		{
			//cout << "tmpExonNum: " << tmp << endl;
			tmpExonLineNO ++;
			int tmpThousandIndex = tmpExonLineNO / 10000;
			if(tmpExonLineNO == tmpThousandIndex * 10000)
			cout << "Tmp Exon #: " << tmpExonLineNO << endl;

			string tmpChrName = chrNameVec_exon[tmp];
			string tmpFeature = featureVec_exon[tmp];
			int tmpStartPos = startPosVec_exon[tmp];
			int tmpEndPos = endPosVec_exon[tmp];
			string tmpStrand = strandVec_exon[tmp];
			string tmpTranscriptId = transcriptIdVec_exon[tmp];
			string tmpTranscriptName = transcriptNameVec_exon[tmp];
			string tmpGeneId = geneIdVec_exon[tmp];
			string tmpGeneName = geneNameVec_exon[tmp];
			
			bool newGene_bool = true;
			for(int tmpGene = 0; tmpGene < geneNameVec.size(); tmpGene++)//check if new gene or not
			{
				string tmpExistGeneName = geneNameVec[tmpGene];
				if(tmpGeneName == tmpExistGeneName) // gene already exists
				{
					// check if new transcript or not
					int tmpGeneTranscriptIsoformNum = (transcriptNameVecVec[tmpGene]).size();
					bool newTranscript_bool = true;
					for(int tmpTranscript = 0; tmpTranscript < tmpGeneTranscriptIsoformNum; tmpTranscript++)
					{
						string tmpExistTranscriptName = (transcriptNameVecVec[tmpGene])[tmpTranscript];
						if(tmpTranscriptName == tmpExistTranscriptName) // transcript already exists
						{
							// add new exon
							((exonPosPairVecVecVec[tmpGene])[tmpTranscript]).push_back(
								pair<int,int>(tmpStartPos, tmpEndPos));
							newTranscript_bool = false;
							break;
						}
					}
					if(newTranscript_bool)
					{
						(transcriptNameVecVec[tmpGene]).push_back(tmpTranscriptName);
						(transcriptIdVecVec[tmpGene]).push_back(tmpTranscriptId);
						vector< pair<int,int> > tmpNewTranscriptExonPosPairVec;
						tmpNewTranscriptExonPosPairVec.push_back(pair<int,int>(tmpStartPos, tmpEndPos));
						(exonPosPairVecVecVec[tmpGene]).push_back(tmpNewTranscriptExonPosPairVec);
					}
					newGene_bool = false;
					break;
				}
			}
			if(newGene_bool)
			{
				geneChrNameVec.push_back(tmpChrName);
				geneNameVec.push_back(tmpGeneName);
				geneIdVec.push_back(tmpGeneId);
				
				vector<string> tmpNewTranscriptNameVec;
				tmpNewTranscriptNameVec.push_back(tmpTranscriptName);
				transcriptNameVecVec.push_back(tmpNewTranscriptNameVec);

				vector<string> tmpNewTranscriptIdVec;
				tmpNewTranscriptIdVec.push_back(tmpTranscriptId);
				transcriptIdVecVec.push_back(tmpNewTranscriptIdVec);

				vector< pair<int,int> > tmpExonPosPairVec;
				tmpExonPosPairVec.push_back(pair<int,int>(tmpStartPos, tmpEndPos));
				vector< vector< pair<int,int> > > tmpNewTranscriptExonPosPairVecVec;
				tmpNewTranscriptExonPosPairVecVec.push_back(tmpExonPosPairVec);
				exonPosPairVecVecVec.push_back(tmpNewTranscriptExonPosPairVecVec);
			}
		}
		cout << "gene #: " << geneNameVec.size() << endl;
		for(int tmp = 0; tmp < geneNameVec.size(); tmp++)
		{
			Gene_Transcript_Info tmpGeneTranscriptInfo;
			tmpGeneTranscriptInfo.initiate(geneChrNameVec[tmp], geneNameVec[tmp], geneIdVec[tmp], 
				transcriptNameVecVec[tmp], transcriptIdVecVec[tmp], exonPosPairVecVecVec[tmp], indexInfo);
			geneTranscriptInfoVec.push_back(tmpGeneTranscriptInfo);
		}
	}

	void parseStr2fieldVec(vector<string>& tmpFieldVec, string& tmpStr)
	{
		int startLoc = 0;
		for(int tmp = 0; ; tmp++)
		{
			int tabLoc = tmpStr.find("\t", startLoc);
			if(tabLoc == string::npos)
				break;
			string tmpField = tmpStr.substr(startLoc, tabLoc-startLoc);
			tmpFieldVec.push_back(tmpField);
			startLoc = tabLoc + 1;
		}
		tmpFieldVec.push_back(tmpStr.substr(startLoc));
	}

	bool parseEnsemblStr(string& tmpStr, string& tmpChrName, string& tmpFeature, 
		int& tmpStartPos, int& tmpEndPos, string& tmpStrand, string& tmpTranscriptId, 
		string& tmpTranscriptName, string& tmpGeneId, string& tmpGeneName)
	{
		vector<string> tmpFieldVec;
		this->parseStr2fieldVec(tmpFieldVec, tmpStr);
		if(tmpFieldVec[2] == "exon")
			return false;
		tmpChrName = tmpFieldVec[0];
		tmpFeature = tmpFieldVec[1];
		string tmpStartPosStr = tmpFieldVec[3];
		string tmpEndPosStr = tmpFieldVec[4];
		tmpStartPos = atoi(tmpStartPosStr.c_str());
		tmpEndPos = atoi(tmpEndPosStr.c_str());
		tmpStrand = tmpFieldVec[6];
		string tmpOtherStr = tmpFieldVec[8];
		int tmpLoc_geneId = tmpOtherStr.find("gene_id");
		int tmpLoc_geneName = tmpOtherStr.find("gene_name");
		int tmpLoc_transcriptId = tmpOtherStr.find("transcript_id");
		int tmpLoc_transcriptName = tmpOtherStr.find("transcript_name");		
		if((tmpLoc_geneId == string::npos)||(tmpLoc_geneName == string::npos)||
			(tmpLoc_transcriptId == string::npos)||(tmpLoc_transcriptName == string::npos))
			return false;
		int quoteLoc_after_geneId = tmpOtherStr.find("\"", tmpLoc_geneId);
		int commaLoc_after_geneId = tmpOtherStr.find(";", tmpLoc_geneId);
		int quoteLoc_after_geneName = tmpOtherStr.find("\"", tmpLoc_geneName);
		int commaLoc_after_geneName = tmpOtherStr.find(";", tmpLoc_geneName);		
		int quoteLoc_after_transcriptId = tmpOtherStr.find("\"", tmpLoc_transcriptId);
		int commaLoc_after_transcriptId = tmpOtherStr.find(";", tmpLoc_transcriptId);
		int quoteLoc_after_transcriptName = tmpOtherStr.find("\"", tmpLoc_transcriptName);
		int commaLoc_after_transcriptName = tmpOtherStr.find(";", tmpLoc_transcriptName);	
		if((quoteLoc_after_geneId == string::npos)||(commaLoc_after_geneId == string::npos)
			||(quoteLoc_after_geneName == string::npos)||(commaLoc_after_geneName == string::npos)
			||(quoteLoc_after_transcriptId == string::npos)||(commaLoc_after_transcriptId == string::npos)
			||(quoteLoc_after_transcriptName == string::npos)||(commaLoc_after_transcriptName == string::npos))
			return false;
		tmpGeneId = tmpOtherStr.substr(quoteLoc_after_geneId + 1, commaLoc_after_geneId - quoteLoc_after_geneId - 2);
		tmpGeneName = tmpOtherStr.substr(quoteLoc_after_geneName + 1, commaLoc_after_geneName - quoteLoc_after_geneName - 2);
		tmpTranscriptId = tmpOtherStr.substr(quoteLoc_after_transcriptId + 1, commaLoc_after_transcriptId - quoteLoc_after_transcriptId - 2);
		tmpTranscriptName = tmpOtherStr.substr(quoteLoc_after_transcriptName + 1, commaLoc_after_transcriptName - quoteLoc_after_transcriptName - 2);		
		return true;
	}

	int returnGeneNum()
	{
		return geneTranscriptInfoVec.size();
	}

	int returnChrNameInt_specificGene(int tmpIndex)
	{
		return geneTranscriptInfoVec[tmpIndex].returnChrNameInt();
	}

	int returnStartPos_specificGene(int tmpIndex)
	{
		return geneTranscriptInfoVec[tmpIndex].returnStartPos();
	}

	int returnEndPos_specificGene(int tmpIndex)
	{
		return geneTranscriptInfoVec[tmpIndex].returnEndPos();
	}

	int returnTranscriptIsoformNum_specificGene(int tmpIndex)
	{
		return geneTranscriptInfoVec[tmpIndex].returnTranscriptIsoformNum();
	}

	void mkDir_printoutFa_gene(string& outputFaDir_gene, Index_Info* indexInfo)
	{
		outputFaDir_gene += "/";
		string cmd_outputFaDir_gene = "mkdir " + outputFaDir_gene;
		system(cmd_outputFaDir_gene.c_str());
		cout << "geneTranscriptInfoVec.size(): " << geneTranscriptInfoVec.size() << endl;
		for(int tmp = 0; tmp < geneTranscriptInfoVec.size(); tmp++)
		{
			string tmpGeneName = geneTranscriptInfoVec[tmp].return_geneName();
			//cout << "tmpGeneName: " << tmpGeneName << endl;
			string tmpFile = outputFaDir_gene + tmpGeneName + "_geneOnly_seq.fa";
			geneTranscriptInfoVec[tmp].printoutFa_gene(tmpFile, indexInfo);
		}
	}

	void mkDir_printoutFa_transcript(string& outputFaDir_transcript, Index_Info* indexInfo)
	{
		outputFaDir_transcript += "/";
		string cmd_outputFaDir_transcript = "mkdir " + outputFaDir_transcript;
		system(cmd_outputFaDir_transcript.c_str());
		for(int tmp = 0; tmp < geneTranscriptInfoVec.size(); tmp++)
		{
			string tmpGeneName = geneTranscriptInfoVec[tmp].return_geneName();
			string tmpFile = outputFaDir_transcript + tmpGeneName + "_transcriptOnly_seq.fa";
			geneTranscriptInfoVec[tmp].printoutFa_transcript(tmpFile, indexInfo);
		}
	}	

	void mkDir_printoutFa_gene_transcript(string& outputFaDir_gene_transcript, Index_Info* indexInfo)
	{
		outputFaDir_gene_transcript += "/";
		string cmd_outputFaDir_gene_transcript = "mkdir " + outputFaDir_gene_transcript;
		system(cmd_outputFaDir_gene_transcript.c_str());
		for(int tmp = 0; tmp < geneTranscriptInfoVec.size(); tmp++)
		{
			string tmpGeneName = geneTranscriptInfoVec[tmp].return_geneName();
			string tmpFile = outputFaDir_gene_transcript + tmpGeneName + "_gene_transcript_seq.fa";
			geneTranscriptInfoVec[tmp].printoutFa_gene_transcript(tmpFile, indexInfo);
		}
	}
};
#endif